--- a/compiler-rt/lib/rtsan/CMakeLists.txt
+++ b/compiler-rt/lib/rtsan/CMakeLists.txt
@@ -28,10 +28,9 @@ set(RTSAN_HEADERS
 set(RTSAN_DEPS)
 
 set(RTSAN_CFLAGS
-  ${COMPILER_RT_COMMON_CFLAGS}
-  ${COMPILER_RT_CXX_CFLAGS}
+  ${SANITIZER_COMMON_CFLAGS}
   -DSANITIZER_COMMON_NO_REDEFINE_BUILTINS)
-set(RTSAN_LINK_FLAGS ${COMPILER_RT_COMMON_LINK_FLAGS})
+set(RTSAN_LINK_FLAGS ${SANITIZER_COMMON_LINK_FLAGS})
 set(RTSAN_DYNAMIC_LIBS
   ${COMPILER_RT_UNWINDER_LINK_LIBS}
   ${SANITIZER_CXX_ABI_LIBRARIES}
--- a/compiler-rt/lib/rtsan/rtsan_context.cpp
+++ b/compiler-rt/lib/rtsan/rtsan_context.cpp
@@ -12,8 +12,8 @@
 #include "rtsan/rtsan.h"
 
 #include "sanitizer_common/sanitizer_allocator_internal.h"
+#include "sanitizer_common/sanitizer_placement_new.h"
 
-#include <new>
 #include <pthread.h>
 
 using namespace __sanitizer;
--- a/compiler-rt/lib/rtsan/rtsan_suppressions.cpp
+++ b/compiler-rt/lib/rtsan/rtsan_suppressions.cpp
@@ -16,11 +16,10 @@
 
 #include "sanitizer_common/sanitizer_common.h"
 #include "sanitizer_common/sanitizer_internal_defs.h"
+#include "sanitizer_common/sanitizer_placement_new.h"
 #include "sanitizer_common/sanitizer_suppressions.h"
 #include "sanitizer_common/sanitizer_symbolizer.h"
 
-#include <new>
-
 using namespace __sanitizer;
 using namespace __rtsan;
 
