--- a/test-release.sh
+++ b/test-release.sh
@@ -737,13 +737,6 @@ for Flavor in $Flavors ; do
         $llvmCore_phase3_objdir $llvmCore_phase3_destdir
     clean_RPATH $llvmCore_phase3_destdir/usr/local
 
-    ########################################################################
-    # Testing: Test phase 3
-    c_compiler=$llvmCore_phase3_destdir/usr/local/bin/clang
-    cxx_compiler=$llvmCore_phase3_destdir/usr/local/bin/clang++
-    echo "# Testing - built with clang"
-    test_llvmCore 3 $Flavor $llvmCore_phase3_objdir
-
     ########################################################################
     # Compare .o files between Phase2 and Phase3 and report which ones
     # differ.
